ml-agents-envs/mlagents/envs/base_env.py

```diff
@@ -0,0 +1,305 @@
+     across simulation steps.
+     - action_mask is an optional list of two dimensional array of booleans.
+     Only available in multi-discrete action space type.
+     Each array corresponds to an action branch. The first dimension of each
+     array is the batch size and the second contains a mask for each action of
+     the branch. If true, the action is not available for the agent during
+     this simulation step.
+    """
+
+    def __init__(self, obs, reward, done, max_step, agent_id, action_mask):
+        self.obs: List[np.array] = obs
+        self.reward: np.array = reward
+        self.done: np.array = done
+        self.max_step: np.array = max_step
+        self.agent_id: np.array = agent_id
+        self.action_mask: Optional[List[np.array]] = action_mask
+        self._agent_id_to_index: Optional[Dict[int, int]] = None
+
+    def contains_agent(self, agent_id: int) -> bool:
+        if self._agent_id_to_index is None:
+            self._agent_id_to_index = {}
+            for a_idx, a_id in enumerate(self.agent_id):
+                self._agent_id_to_index[a_id] = a_idx
+        return agent_id in self._agent_id_to_index
+
+    def get_agent_step_result(self, agent_id: int) -> StepResult:
+        """
+        returns the step result for a specific agent.
+        :param agent_id: The id of the agent
+        :returns: obs, reward, done, agent_id and optional action mask for a
+        specific agent
+        """
+        if not self.contains_agent(agent_id):
+            raise IndexError(
+                "agent_id {} is not present in the BatchedStepResult".format(agent_id)
+            )
+        agent_index = self._agent_id_to_index[agent_id]  # type: ignore
+        agent_obs = []
+        for batched_obs in self.obs:
+            agent_obs.append(batched_obs[agent_index])
+        agent_mask = None
+        if self.action_mask is not None:
+            agent_mask = []
+            for mask in self.action_mask:
+                agent_mask.append(mask[agent_index])
+        return StepResult(
+            obs=agent_obs,
+            reward=self.reward[agent_index],
+            done=self.done[agent_index],
+            max_step=self.max_step[agent_index],
+            agent_id=agent_id,
+            action_mask=agent_mask,
+        )
+
+    @staticmethod
+    def empty(spec: "AgentGroupSpec") -> "BatchedStepResult":
+        """
+        Returns an empty BatchedStepResult.
+        :param spec: The AgentGroupSpec for the BatchedStepResult
+        """
+        obs: List[np.array] = []
+        for shape in spec.observation_shapes:
+            obs += [np.zeros((0,) + shape, dtype=np.float32)]
+        return BatchedStepResult(
+            obs=obs,
+            reward=np.zeros(0, dtype=np.float32),
+            done=np.zeros(0, dtype=np.bool),
+            max_step=np.zeros(0, dtype=np.bool),
+            agent_id=np.zeros(0, dtype=np.int32),
+            action_mask=None,
+        )
+
+    def n_agents(self) -> int:
```

Comment:
Can you take the length of a numpy array?

Reply:
Looks like it : 
```
>>> import numpy as np
>>> a = np.array([[1,2,3],[4,5,6]])
>>> len(a)
2
>>> 
```

Reply:
I see, they implemented `__len__` as the size of the first dimension. It's probably safer to use `size` but either will work.

Reply:
size will give the product of all dimensions, len will give the first one. I could use .shape[0] but it does not make a huge difference to me.

Reply:
len is fine